### DISCUSSION TITLE
Why does there need to be this loop in the reaper?

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#129] [Q&A] [answered]
Under what circumstances is more than one pass needed in this loop in the reaper?
https://github.com/congo-cc/congo-parser-generator/blob/3018a81635557a035c4522aa9596bf8969ed6d2a/src/java/org/congocc/codegen/java/Reaper.java#L33-L40

**Comment by revusky** [comment, marked as answer]
Well, if you think about the case of a private method. Unless some other non-private code references it -- directly or indirectly -- it can be reaped, right? Of course, once you establish that this method is not to be reaped, then all of the methods/variables that it references also should not be reaped. So, if you figure out in one pass that the method should not be reaped, in the next pass, you are going to add all of the things it references to the list of things not to be reaped. When you do a pass and don't add anything to the list, then you stop. But it typically takes several passes to build up the lists of used methods/types/fields. I don't honestly know whether this is the typical algorithm or the best one. It does seem to work okay. And actually, the reaper is pretty fast.

**Reply by vsajip** [reply]
Ah, of course. I was thinking with my Python hat on, where private methods aren't really a thing.
